shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java

```diff
@@ -84,22 +90,46 @@ public DartMessengerTaskQueue makeBackgroundTaskQueue() {
     }
   }
 
-  private static class DefaultTaskQueue implements DartMessengerTaskQueue {
+  /** A serial task queue that can run on a concurrent ExecutorService. */
+  static class DefaultTaskQueue implements DartMessengerTaskQueue {
     @NonNull private final ExecutorService executor;
+    @NonNull private final ConcurrentLinkedQueue<Runnable> queue;
+    @NonNull private final AtomicBoolean isRunning;
 
-    DefaultTaskQueue() {
-      // TODO(gaaclarke): Use a shared thread pool with serial queues instead of
-      // making a thread for each TaskQueue.
-      ThreadFactory threadFactory =
-          (Runnable runnable) -> {
-            return new Thread(runnable, "DartMessenger.DefaultTaskQueue");
-          };
-      this.executor = Executors.newSingleThreadExecutor(threadFactory);
+    DefaultTaskQueue(ExecutorService executor) {
+      this.executor = executor;
+      queue = new ConcurrentLinkedQueue<>();
+      isRunning = new AtomicBoolean(false);
     }
 
     @Override
     public void dispatch(@NonNull Runnable runnable) {
-      executor.execute(runnable);
+      queue.add(runnable);
+      executor.execute(
+          () -> {
+            flush();
+          });
+    }
+
+    private void flush() {
+      // Don't execute if we are already executing (enforce serial execution).
+      if (isRunning.compareAndSet(false, true)) {
+        try {
+          @Nullable Runnable runnable = queue.poll();
+          if (runnable != null) {
+            runnable.run();
```

Comment:
Why not just flush the whole queue from here? What's the advantage of resubmitting to the executor for each task?

Reply:
We could.  This is more fair to allow other tasks to execute so one TaskQueue can't dominate a whole thread.

Reply:
Ok. I think that makes sense. This may be something to revisit as we look at more traces/benchmarks - particularly in cases on devices with few cores.